source/Halibut.Tests/Transport/Streams/TimeoutStreamWrapperFixture.cs

```diff
@@ -0,0 +1,369 @@
+                    {
+                        try
+                        {
+                            await sut.ReadAsync(new byte[19], 0, 19, readTokenSource.Token);
+                        }
+                        catch (Exception e)
+                        {
+                            actualException = e;
+                            break;
+                        }
+                    }
+                    stopWatch.Stop();
+
+                    actualException.Should().NotBeNull().And.BeOfType<OperationCanceledException>();
+                    actualException!.Message.Should().Be("The ReadAsync operation was cancelled.");
+
+                    stopWatch.Elapsed.Should().BeLessThan(TimeSpan.FromSeconds(10));
+                }
+            }
+        }
+
+        [Test]
+        public async Task ReadAsyncShouldTimeout()
+        {
+            var (disposables, sut, _) = await BuildTcpClientAndTcpListener(
+                CancellationToken, 
+                onListenerRead: async _ => await Task.Delay(TimeSpan.FromDays(1), CancellationToken));
+
+            using (disposables)
+            {
+                // Ensure the correct timeout is used
+                sut.WriteTimeout = (int)TimeSpan.FromSeconds(120).TotalMilliseconds;
+                sut.ReadTimeout = (int)TimeSpan.FromSeconds(5).TotalMilliseconds;
+
+                var stopWatch = Stopwatch.StartNew();
+                Exception? actualException = null;
+
+                // Brute force attempt to get the Write to be slow
+                while (!CancellationToken.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await sut.ReadAsync(new byte[19], 0, 19, CancellationToken);
+                    }
+                    catch (Exception e)
+                    {
+                        actualException = e;
+                        break;
+                    }
+                }
+
+                stopWatch.Stop();
+
+                actualException.Should().NotBeNull().And.BeOfType<OperationCanceledException>();
+                actualException!.Message.Should().Be("The ReadAsync operation timed out after 00:00:05.");
+
+                stopWatch.Elapsed.Should().BeLessThan(TimeSpan.FromSeconds(10));
+            }
+        }
+
+        [Test]
+        public async Task WriteShouldPassThrough()
+        {
+            string? readData = null;
+            
+            var (disposables, sut, _) = await BuildTcpClientAndTcpListener(
+                CancellationToken,
+                onListenerRead: async data =>
+                {
+                    await Task.CompletedTask;
+                    readData += data;
+                });
+
+            using (disposables)
+            {
+                var buffer = Encoding.UTF8.GetBytes("Test");
+                sut.Write(buffer, 0, buffer.Length);
+
+                while (readData == null && !CancellationToken.IsCancellationRequested)
```

Comment:
Is it guaranteed to read all the data in one block?

If so, should the onListenerRead be `readData = data`?

If not, then would this test not be a risk of being flaky with the while loop only checking for `readData == null`?

Reply:
We have a Try.CatchingError helper that I will adapt to async and use here to remove boilerplate

Reply:
I don't know if its guaranteed so I've made it more robust in terms of the length of data it's waiting for